src/utils.rs

```diff
@@ -202,4 +233,15 @@ mod test {
             assert_eq!(x, vec![1u8; i]);
         }
     }
+
+    #[test]
+    fn test_mlock_munlock() {
+        let t = b"hello world";
+        let mut x = Vec::new();
+        x.extend_from_slice(t);
+        assert!(mlock(&mut x).is_ok());
```

Comment:
How about tests for `is_err`?

Reply:
Both `sodium_mlock()` and `sodium_munlock()` are platform dependent functions whose failure is operating system and local configuration dependent. I expect most of the error conditions are not possible with rust allocated `&[u8]` and triggering the system or per-process limit for locked memory probably needs complicated test setup for setting limits.

If there's error conditions which we could trigger, I'm more than happy to try implementing those.

Ref:
 * https://github.com/jedisct1/libsodium/blob/a6ef940634c6191ceb36a73f8c355655d4d6c740/src/libsodium/sodium/utils.c#L406-L420
 * https://github.com/jedisct1/libsodium/blob/a6ef940634c6191ceb36a73f8c355655d4d6c740/src/libsodium/sodium/utils.c#L422-L437

Reply:
How about mlocking it twice or trying to munlock a non-mlocked memory region?

Reply:
double locking or unlocking not-locked memory in macOS work fine. No errors there. Reading the POSIX IEEE Std 1003.1-2017 and man pages from Linux, OpenBSD and macOS seem to not document such error conditions.

Also from Linux man page:

> Memory locks do not stack, that is, pages which have been locked several times by calls to mlock(), mlock2(), or mlockall() will be unlocked by a single call to munlock() for the corresponding range or by munlockall().

Ref. http://man7.org/linux/man-pages/man2/mlock.2.html

Reply:
Ah. How do we test it then?

Reply:
I added a unix specific test which uses [`setrlimit(2)`](https://linux.die.net/man/2/setrlimit) to set a limit for locked memory size which then causes `mlock()` to fail. I tested this with Linux and macOS. See https://github.com/sodiumoxide/sodiumoxide/pull/324/commits/68f5a64b32e48c1bae3bbc26fc568448d8ea56e9.

However, this feels a bit wrong since it's success depends on compile time parameters of libsodium itself. If libsodium gets compiled without mlock() support, then this test fails. See https://github.com/jedisct1/libsodium/blob/a6ef940634c6191ceb36a73f8c355655d4d6c740/src/libsodium/sodium/utils.c#L412-L413